packages/aggregator/src/analytics/aggregateAnalytics/aggregations/getFinalValuePerPeriod.js

```diff
@@ -106,28 +107,60 @@ class FinalValueAggregator {
     this.cache = cache;
   }
 
-  getContinuousValues(analytics, aggregationPeriod) {
-    const periods = getContinuousPeriodsForAnalytics(analytics, aggregationPeriod, true);
+  getContinuousValues(organisationUnitCache, periods) {
+    let mostRecentValue;
 
+    return periods.map(period => {
+      mostRecentValue = organisationUnitCache[period] || mostRecentValue;
+      return {
+        ...mostRecentValue,
+        period,
+      };
+    });
+  }
+
+  fillWithPresetValues(organisationUnitCache, periods, value) {
     const values = [];
-    this.cache.iterateOrganisationUnitCache(organisationUnitCache => {
-      let mostRecentValue;
-
-      periods.forEach(period => {
-        const valueForPeriod = organisationUnitCache[period];
-
-        if (valueForPeriod !== undefined) {
-          mostRecentValue = valueForPeriod;
-        }
-        if (mostRecentValue !== undefined) {
-          values.push({
-            ...mostRecentValue,
-            period,
+    const dataElements = uniq(Object.values(organisationUnitCache).map(data => data.dataElement));
+    const organisationUnits = uniq(
+      Object.values(organisationUnitCache).map(data => data.organisationUnit),
+    );
+
+    periods.forEach(period => {
+      const valueForPeriodFromCache = organisationUnitCache[period];
+      if (valueForPeriodFromCache) {
+        values.push(valueForPeriodFromCache);
+      } else {
+        // Assign empty period with value for each dataElement mapping each orgUnit that existed in results.
+        dataElements.forEach(dataElement => {
+          organisationUnits.forEach(organisationUnit => {
+            values.push({
+              dataElement,
+              organisationUnit,
+              value,
+              period,
+            });
           });
-        }
-      });
+        });
+      }
     });
+    return values;
+  }
+
+  getFilledValues(analytics, aggregationPeriod, fillEmptyPeriodsWith) {
+    const checkIfFillingContinuousValues = fillEmptyPeriodsWith === 'previous';
+    const periods = getContinuousPeriodsForAnalytics(
+      analytics,
+      aggregationPeriod,
+      checkIfFillingContinuousValues,
+    );
 
+    let values;
+    this.cache.iterateOrganisationUnitCache(organisationUnitCache => {
+      values = checkIfFillingContinuousValues
+        ? this.getContinuousValues(organisationUnitCache, periods)
+        : this.fillWithPresetValues(organisationUnitCache, periods, fillEmptyPeriodsWith);
+    });
```

Comment:
Hmmm I think something isn't working here, on my local I see incorrect data for a COVID AU dashboard that uses this aggregation:
![image](https://user-images.githubusercontent.com/59544282/116853137-e5e3f500-ac38-11eb-8503-86a6307adb2e.png)


Tbh I think we can take a step back here and use much simpler logic based on the original code to handle this logic. The logic should be:
- Where we were using the `mostRecentValue`, now use the `mostRecentValue` if `fillEmptyPeriodsWith` = `previous` otherwise use `fillEmptyPeriodsWith`

Reply:
This one works very well with less changes. Nice one!